system/lib/pthread/library_pthread.c

```diff
@@ -457,11 +540,73 @@ void EMSCRIPTEN_KEEPALIVE emscripten_main_thread_process_queued_calls()
 	// This must be before pthread_mutex_lock(), since pthread_mutex_lock() can call back to this function.
 	bool_inside_nested_process_queued_calls = 1;
 	pthread_mutex_lock(&call_queue_lock);
-	for (int i = 0; i < call_queue_length; ++i)
-		_do_call(call_queue[i]);
-	call_queue_length = 0;
-	bool_inside_nested_process_queued_calls = 0;
+	int head = emscripten_atomic_load_u32((void*)&call_queue_head);
+	int tail = emscripten_atomic_load_u32((void*)&call_queue_tail);
+	while (head != tail)
+	{
+		// Assume that the call is heavy, so unlock access to the call queue while it is being performed.
```

Comment:
I thinks we should also unlock `bool_inside_nested_process_queued_calls` during the call, to minimize possibility of deadlocks. The important part is to prevent reentry on `pthread_mutex_lock` here, but the call itself should be fine, as long as it has been removed from the queue. That's how I did it in the original PR #4325.